File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/express/offline_methods/OfflineMethodsAdapter.java
package com.mercadopago.android.px.internal.features.express.offline_methods;

import android.annotation.SuppressLint;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.mercadopago.android.px.R;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/* default */ final class OfflineMethodsAdapter extends RecyclerView.Adapter<OfflineMethodsRowHolder> {

    private final List<OfflineMethodItem> offlineItems = new ArrayList<>();
    private OfflineMethodsRowHolder lastHolder;
    private int lastPositionSelected = -1;
    private final OfflineMethodsFragment.OnMethodSelectedListener onMethodSelectedListener;

    public OfflineMethodsAdapter(final OfflineMethodsFragment.OnMethodSelectedListener onMethodSelectedListener) {
        this.onMethodSelectedListener = onMethodSelectedListener;
    }

    @NonNull
    @Override
    public OfflineMethodsRowHolder onCreateViewHolder(@NonNull final ViewGroup parent, final int i) {
        final View offlineMethodView = LayoutInflater.from(parent.getContext())
            .inflate(R.layout.px_view_offline_item, parent, false);

        return new OfflineMethodsRowHolder(offlineMethodView);
    }

    @Override
    public void onBindViewHolder(@NonNull final OfflineMethodsRowHolder holder,
        @SuppressLint("RecyclerView") final int position) {
        holder.setChecked(lastPositionSelected >= 0 && lastPositionSelected == position);

        holder.populate(offlineItems.get(position), () -> {
            if (lastHolder != null) {
                lastHolder.setChecked(false);
            }

            holder.setChecked(true);
            lastHolder = holder;
            lastPositionSelected = position;

            onMethodSelectedListener.onItemSelected(offlineItems.get(position));
        });
    }

    public void setItems(@NonNull final Collection<OfflineMethodItem> items) {
        offlineItems.clear();
        offlineItems.addAll(items);
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return offlineItems.size();
    }

    interface OnItemClicked {
        void onClick();
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/express/offline_methods/OfflineMethodsAdapter.kt
package com.mercadopago.android.px.internal.features.express.offline_methods

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.mercadopago.android.px.R
import java.util.*

internal class OfflineMethodsAdapter(private val onMethodSelectedListener: OfflineMethods.OnMethodSelectedListener)
    : RecyclerView.Adapter<OfflineMethodsRowHolder>() {

    private val offlineItems: MutableList<OfflineMethodItem> = ArrayList()
    private var lastHolder: OfflineMethodsRowHolder? = null
    private var lastPositionSelected = -1

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): OfflineMethodsRowHolder {
        val offlineMethodView = LayoutInflater.from(parent.context)
            .inflate(R.layout.px_view_offline_item, parent, false)
        return OfflineMethodsRowHolder(offlineMethodView)
    }

    override fun onBindViewHolder(holder: OfflineMethodsRowHolder, position: Int) {
        holder.setChecked(lastPositionSelected >= 0 && lastPositionSelected == position)
        holder.populate(offlineItems[position], object : OnItemClicked {
            override fun onClick() {
                lastHolder?.setChecked(false)
                holder.setChecked(true)
                lastHolder = holder
                lastPositionSelected = position
                onMethodSelectedListener.onItemSelected(offlineItems[position])
            }
        })
    }

    fun setItems(items: Collection<OfflineMethodItem>) {
        offlineItems.clear()
        offlineItems.addAll(items)
        offlineItems.firstOrNull { item -> !item.isOfflinePaymentTypeItem }?.run {
            lastPositionSelected = offlineItems.indexOf(this)
            onMethodSelectedListener.onItemSelected(this)
        }
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return offlineItems.size
    }

    internal interface OnItemClicked {
        fun onClick()
    }

}